src/manifolds/GeneralizedStiefel.jl

```diff
@@ -0,0 +1,223 @@
+        )
+    end
+    if any(size(p) != representation_size(M))
+        return DomainError(
+            size(p),
+            "The matrix $(p) is does not lie on the Generalized Stiefel manifold of dimension ($(n),$(k)), since its dimensions are wrong.",
+        )
+    end
+    c = p' * M.B * p
+    if !isapprox(c, one(c); kwargs...)
+        return DomainError(
+            norm(c - one(c)),
+            "The point $(p) does not lie on the Generalized Stiefel manifold of dimension ($(n),$(k)), because x'Bx is not the unit matrix.",
+        )
+    end
+end
+
+
+@doc doc"""
+    check_tangent_vector(M::GeneralizedStiefel, p, X; kwargs...)
+
+Check whether `X` is a valid tangent vector at `p` on the [`GeneralizedStiefel`](@ref)
+`M`=$\operatorname{St}(n,k,B)$, i.e. the [`AbstractNumbers`](@ref) fits,
+`p` is a valid point on `M` and
+it (approximately) holds that $p^{\mathrm{H}}BX + X^{\mathrm{H}}Bp = 0$, where
+`kwargs...` is passed to the `isapprox`.
+"""
+function check_tangent_vector(M::GeneralizedStiefel{n,k,T}, p, X; kwargs...) where {n,k,T}
+    mpe = check_manifold_point(M, p, kwargs)
+    mpe === nothing || return mpe
+    if (T === ℝ) && !(eltype(X) <: Real)
+        return DomainError(
+            eltype(X),
+            "The matrix $(X) is not a real-valued matrix, so it can not be a tangent vector to the Generalized Stiefel manifold of dimension ($(n),$(k)).",
+        )
+    end
+    if (T === ℂ) && !(eltype(X) <: Real) && !(eltype(v) <: Complex)
+        return DomainError(
+            eltype(X),
+            "The matrix $(X) is a neither real- nor complex-valued matrix, so it can not be a tangent vector to the complex Generalized Stiefel manifold of dimension ($(n),$(k)).",
+        )
+    end
+    if any(size(X) != representation_size(M))
+        return DomainError(
+            size(X),
+            "The matrix $(X) does not lie in the tangent space of $(p) on the Generalized Stiefel manifold of dimension ($(n),$(k)), since its dimensions are wrong.",
+        )
+    end
+    if !isapprox(p' * M.B * X + X' * M.B * p, zeros(k, k); kwargs...)
+        return DomainError(
+            norm(p' * M.B * X + X' * M.B * p),
+            "The matrix $(X) does not lie in the tangent space of $(p) on the Generalized Stiefel manifold of dimension ($(n),$(k)), since x'Bv + v'Bx is not the zero matrix.",
+        )
+    end
+end
+
+embed!(::GeneralizedStiefel, q, p) = (q .= p)
+
+embed!(::GeneralizedStiefel, Y, p, X) = (Y .= X)
+
+@doc doc"""
+    inner(M::GeneralizedStiefel, p, X, Y)
+
+Compute the inner product for two tangent vectors `X`, `Y` from the tangent space of `p`
+on the [`GeneralizedStiefel`](@ref) manifold `M`. The formula reads
+
+````math
+(X, Y)_p = \operatorname{trace}(v^{\mathrm{H}}Bw),
+````
+i.e. the metric induced by the scalar product `B` from the embedding, restricted to the
+tangent space.
+"""
+inner(M::GeneralizedStiefel, p, X, Y) = dot(X, M.B * Y)
```

Comment:
I think it doesn't look like an isometric embedding when we have a custom `inner`?

```julia
struct GeneralizedStiefel{n,k,F,TB<:AbstractMatrix} <: AbstractEmbeddedManifold{AbstractIsometricEmbeddingType}
```

Reply:
Good that you noticed, that was too  fast copied from Stiefel where this is indeed the case (and this PR also covers transferring Stiefel to exactely this).